File: services/paper_service.py
from models.paper import Paper
from extensions import db
from sqlalchemy import or_

def add_paper(data):
    paper = Paper(
        course=data.get('course'),
        year=data.get('year'),
        term=data.get('term'),
        college=data.get('college'),
        image_path=data.get('image_path')
    )
    db.session.add(paper)
    db.session.commit()
    return paper

def update_paper_image(paper_id, image_path):
    paper = Paper.query.get(paper_id)
    if paper:
        paper.image_path = image_path
        db.session.commit()
        return paper
    return None

def get_papers(course=None, year=None, term=None, college=None, keyword=None, page=1, per_page=10):
    query = Paper.query
    if course:
        query = query.filter(Paper.course.like(f'%{course}%'))
    if year:
        query = query.filter_by(year=year)
    if term:
        query = query.filter(Paper.term.like(f'%{term}%'))
    if college:
        query = query.filter(Paper.college.like(f'%{college}%'))
    if keyword:
        query = query.filter(
            or_(
                Paper.course.like(f'%{keyword}%'),
                Paper.term.like(f'%{keyword}%'),
                Paper.college.like(f'%{keyword}%')
            )
        )
    pagination = query.order_by(Paper.created_at.desc()).paginate(page=page, per_page=per_page, error_out=False)
    return pagination 

File: backend-go/services/paper_service.go
package services

import (
	"backend-go/database"
	"backend-go/models"
)

type PaperInput struct {
	Course    string `json:"course"`
	Year      int    `json:"year"`
	Term      string `json:"term"`
	College   string `json:"college"`
	ImagePath string `json:"image_path"`
	FileType  string `json:"file_type"`
	FileSize  int    `json:"file_size"`
	FileURL   string `json:"file_url"`
}

func AddPaper(input PaperInput) (*models.Paper, error) {
	paper := models.Paper{
		Course:    input.Course,
		Year:      input.Year,
		Term:      input.Term,
		College:   input.College,
		ImagePath: input.ImagePath,
	}

	if err := database.DB.Create(&paper).Error; err != nil {
		return nil, err
	}

	return &paper, nil
}

func UpdatePaperImage(paperID uint, imagePath string) (*models.Paper, error) {
	var paper models.Paper
	if err := database.DB.First(&paper, paperID).Error; err != nil {
		return nil, err
	}

	paper.ImagePath = imagePath
	if err := database.DB.Save(&paper).Error; err != nil {
		return nil, err
	}

	return &paper, nil
}

func GetPapers(course, term, college, keyword string, year, page, perPage int) ([]models.Paper, int64, error) {
	var papers []models.Paper
	var total int64

	query := database.DB.Model(&models.Paper{})

	if course != "" {
		query = query.Where("course LIKE ?", "%"+course+"%")
	}
	if year > 0 {
		query = query.Where("year = ?", year)
	}
	if term != "" {
		query = query.Where("term LIKE ?", "%"+term+"%")
	}
	if college != "" {
		query = query.Where("college LIKE ?", "%"+college+"%")
	}
	if keyword != "" {
		query = query.Where("course LIKE ? OR term LIKE ? OR college LIKE ?",
			"%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%")
	}

	query.Count(&total)

	offset := (page - 1) * perPage
	if err := query.Order("created_at DESC").Offset(offset).Limit(perPage).Find(&papers).Error; err != nil {
		return nil, 0, err
	}

	return papers, total, nil
} 